File: go/day1part2.go
package main

/*
aoc2022 - Day 1, Part 2
-----------------------
Hungry Hungry Elves

Each  Elf enumerates the calorific content of their snacks, 1 snack per line.
Elves are separated in the input by a blank line.

Find the top THREE chonk-carrying Elves. How many calories do they carry all together?
*/

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

// for testing:
//var dataFile = "../data/day1test.txt"

// for Stars:
var dataFile = "../data/day1input.txt"

func main() {
	file, err := os.Open(dataFile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	s := bufio.NewScanner(file)

	//Given each LINE is either blank (delimiting elves) or
	//numeric, the obvious thing to do here is build a list
	//incrementing the value by calories per line. However,
	//traversal is a lot easier if we just build a Map
	elfMap := map[int]int{}
	ndx := 1

	for s.Scan() {
		w := s.Text()
		n, err := strconv.Atoi(w)
		if err != nil {
			//couldn't convert str to int, must be new elf
			ndx++
		} else {
			elfMap[ndx] += n
		}
	}

	//Populate the top 3 chonkiest carriers
	//(this is easiest to do by sorting the input by VALUE instead of KEY and taking the last 3 entries)
	elfNdx := make([]int, 0, len(elfMap))
	for e := range elfMap {
		elfNdx = append(elfNdx, e)
	}

	//Sort the elfNdx by the number of calories not the index:
	sort.SliceStable(elfNdx, func(i, j int) bool {
		return elfMap[elfNdx[i]] > elfMap[elfNdx[j]]
	})

	//Now we just need a slice of the first 3 elements in this sorted elfNdx
	var chonks []int = elfNdx[0:3]

	totCals := 0
	for _, e := range chonks {
		fmt.Println(e, elfMap[e])
		totCals += elfMap[e]
	}
	fmt.Println(totCals)

}

File: nodejs/day1part2.js
#!/usr/bin/env node

/*
aoc2022 - Day 1, Part 2
-----------------------
Hungry Hungry Elves

Each  Elf enumerates the calorific content of their snacks, 1 snack per line.
Elves are separated in the input by a blank line.

Find the top THREE chonk-carrying Elves. How many calories do they carry all together?
*/
const fs = require('fs');
const readline = require('readline');

// for testing:
//const dataFile = "data/day1test.txt"
// for Stars:
const dataFile = "data/day1input.txt"

var calorieMap = {};
var elfNdx = 0;

async function main() {
    const rl = readline.createInterface({
        input: fs.createReadStream(dataFile),
        crlfDelay: Infinity
    });

    
    
    for await (const line of rl) {
        const text = line.split(' ')[0];
        v = Number(text)
        if(typeof v === "number" && v > 0) {
            if(typeof calorieMap[elfNdx] === "number") { 
                calorieMap[elfNdx] = (calorieMap[elfNdx] + v);
            } else {
                calorieMap[elfNdx] = (v)
            }
        } else {
            elfNdx = elfNdx + 1
        }
    }

    //post-processing, sort calorieMap descending, slice top 3
    //(convert to Array to facilitate; we don't mind losing elfNum)
    const chonkValues = Object.keys(calorieMap).map(function (e) {
        return calorieMap[e]
    })
    
    chonkValues.sort(function (a,b){ return a - b }); //WTF is default sort "convert to UTF-16 and compare strings" for?
    chonkValues.reverse();
    //Get slice of first 3 values and sum
    var sumCals = 0
    chonkValues.slice(0,3).forEach(function (v) {
        sumCals += v;
    })
    
    console.log(sumCals)
}


if (require.main === module) {
    main();
}
